staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go

```diff
@@ -59,10 +59,22 @@ func doTransformObject(ctx context.Context, obj runtime.Object, opts interface{}
 	if _, ok := obj.(*metav1.Status); ok {
 		return obj, nil
 	}
-	if err := ensureNonNilItems(obj); err != nil {
-		return nil, err
+
+	// ensure that for empty lists we don't return <nil> items.
+	// This is safe to modify without deep-copying the object, as
+	// List objects themselves are never cached.
+	if meta.IsListType(obj) && meta.LenList(obj) == 0 {
+		if err := meta.SetList(obj, []runtime.Object{}); err != nil {
+			return nil, err
+		}
 	}
 
+	// The below conversions also don't require deep-copying the object:
+	// - PartialObjectMetadata is just accessing the data and referencing from
+	//   a new object
+	// - PartialObjectMetadataList is similar to the above
+	// - Table is also just referencing data from a new Object without modifying
```

Comment:
asTable converts incoming objects to outgoing versions using `scope.Convertor.ConvertToVersion` ... is it the safe or unsafe convertor (the one that deepcopies the input)?

Reply:
It's neither of those - it's a dedicate table converter. It's basically this one:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/registry/rest/table.go#L44

Reply:
OK - that's what you meant. This definitely can be unsafe one.

Reply:
OK, given that currently we're not strictly relying on this being non-mutating, let me mostly revert this commit (as we actually deep-copy object before any encoding anyway).